File: gek-core/src/main/java/xyz/fslabo/common/invoke/Invoker.java
package xyz.fslabo.common.invoke;

import xyz.fslabo.annotations.Nullable;

import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

/**
 * This interface represents an invocation handle to executable content, which can be a method, a constructor, a block
 * of code, or any other executable object.
 *
 * @author fredsuvn
 */
@FunctionalInterface
public interface Invoker {

    /**
     * Returns an {@link Invoker} instance for given method by reflecting.
     * <p>
     * Note if underlying exception is wrapped by {@link InvocationTargetException} when calls
     * {@link #invoke(Object, Object...)}, the {@link InvocationTargetException} will be erased, actual cause from
     * {@link InvocationTargetException#getCause()} will be wrapped by {@link InvocationException} then thrown.
     *
     * @param method given method
     * @return {@link Invoker} instance
     */
    static Invoker reflect(Method method) {
        return new JieInvoke.OfMethod(method);
    }

    /**
     * Returns an {@link Invoker} instance for given constructor by reflecting.
     * <p>
     * Note if underlying exception is wrapped by {@link InvocationTargetException} when calls
     * {@link #invoke(Object, Object...)}, the {@link InvocationTargetException} will be erased, actual cause from
     * {@link InvocationTargetException#getCause()} will be wrapped by {@link InvocationException} then thrown.
     *
     * @param constructor given constructor
     * @return {@link Invoker} instance
     */
    static Invoker reflect(Constructor<?> constructor) {
        return new JieInvoke.OfConstructor(constructor);
    }

    /**
     * Returns {@link Invoker} instance for given method by {@link MethodHandles.Lookup#unreflect(Method)}.
     *
     * @param method given method
     * @return {@link Invoker} instance
     */
    static Invoker handle(Method method) {
        return new JieInvoke.OfMethodHandle(method);
    }

    /**
     * Returns an {@link Invoker} instance for given constructor by
     * {@link MethodHandles.Lookup#unreflectConstructor(Constructor)}.
     *
     * @param constructor given constructor
     * @return {@link Invoker} instance
     */
    static Invoker handle(Constructor<?> constructor) {
        return new JieInvoke.OfMethodHandle(constructor);
    }

    /**
     * Returns an {@link Invoker} instance with specified {@link MethodHandle}.
     *
     * @param handle   specified {@link MethodHandle}
     * @param isStatic whether treats the handle as static when invoking
     * @return {@link Invoker} instance
     */
    static Invoker handle(MethodHandle handle, boolean isStatic) {
        return new JieInvoke.OfMethodHandle(handle, isStatic);
    }

    /**
     * Invokes with specified object and arguments.
     * <p>
     * If this invoker references a {@link Method}, {@link Constructor} or {@link MethodHandle}, the usage of its
     * parameters are equivalent to {@link Method#invoke(Object, Object...)}, {@link Constructor#newInstance(Object...)}
     * or {@link MethodHandle#invokeWithArguments(Object...)}. Otherwise, usage of the parameters is up to the
     * implementation.
     * <p>
     * This method only throws {@link InvocationException} if any problem occurs, use
     * {@link InvocationException#getCause()} to get underlying cause.
     *
     * @param obj  specified object
     * @param args specified arguments
     * @return result of invoking
     * @throws InvocationException wraps if any problem occurs
     */
    @Nullable
    Object invoke(@Nullable Object obj, Object... args) throws InvocationException;
}

File: boat-core/src/main/kotlin/xyz/srclab/common/invoke/Invoker.kt
package xyz.srclab.common.invoke

import java.lang.reflect.Constructor
import java.lang.reflect.Method

/**
 * Represents a invokable instance.
 *
 * @see [InvokerGenerator]
 * @see [ReflectInvokerGenerator]
 * @see [MethodHandlerInvokerGenerator]
 */
interface Invoker {

    /**
     * Does invoke, not forcibly.
     */
    fun <T> invoke(obj: Any?, vararg args: Any?): T {
        return doInvoke(obj, false, *args)
    }

    /**
     * Forcibly invokes.
     *
     * @see invoke
     */
    fun <T> enforce(obj: Any?, vararg args: Any?): T {
        return doInvoke(obj, true, *args)
    }

    /**
     * Does invoke with [obj] -- specifies invoking object, and [force] -- specifies whether forcibly invoke.
     */
    fun <T> doInvoke(obj: Any?, force: Boolean, vararg args: Any?): T

    /**
     * Prepares an invoking, not forcibly.
     */
    fun prepareFor(obj: Any?): Invoke {
        return prepareFor(obj, false)
    }

    /**
     * Prepares an invoking.
     */
    fun prepareFor(obj: Any?, force: Boolean): Invoke {
        return object : Invoke {
            override fun <T> start(vararg args: Any?): T {
                return doInvoke(obj, force, *args)
            }
        }
    }

    companion object {

        @JvmField
        val DEFAULT_INVOKER_GENERATOR = ReflectInvokerGenerator

        @JvmName("ofMethod")
        @JvmStatic
        fun Method.toInvoker(): Invoker {
            return DEFAULT_INVOKER_GENERATOR.ofMethod(this)
        }

        @JvmStatic
        fun ofMethod(clazz: Class<*>, methodName: String, vararg parameterTypes: Class<*>): Invoker {
            return DEFAULT_INVOKER_GENERATOR.ofMethod(clazz, methodName, *parameterTypes)
        }

        @JvmName("ofConstructor")
        @JvmStatic
        fun Constructor<*>.toInvoker(): Invoker {
            return DEFAULT_INVOKER_GENERATOR.ofConstructor(this)
        }

        @JvmStatic
        fun ofConstructor(clazz: Class<*>, vararg parameterTypes: Class<*>): Invoker {
            return DEFAULT_INVOKER_GENERATOR.ofConstructor(clazz, *parameterTypes)
        }
    }
}